### ISSUE TITLE
List index out of bounds error on fflib_QueryFactory

### ISSUE THREAD

**Issue by wimvelzeboer** [apex-enterprise-patterns/fflib-apex-common#233]
Hi,
In some cases the fflib_QueryFactory will return a "List index out of bounds" error, thrown by the getFieldPath method when resolving a field on a related object e.g. `.selectField("Profile.Name")`, in combination with the users profile type and a combination of with and without sharing. It works perfect for Standard Salesforce License users.

This is a **unit test** which will replicate the issue:
```@IsTest
private with sharing class fflib_SObjectSelectorTest 
{
	@IsTest
	static void testElevatedSelector()
	{
		Id communityProfileId = [SELECT Id FROM Profile WHERE Name = 'Company Communities User' LIMIT 1].Id;
		User communityUser = new User(
				LastName = 'Smith',
				UserName = 'John.Smith@unittest.salesforce.com',
				Alias = 'JSmith',
				Email = 'John.Smith@unittest.salesforce.com',
				TimeZoneSidKey = 'America/Los_Angeles',
				LocaleSidKey = 'en_US',
				EmailEncodingKey = 'ISO-8859-1',
				LanguageLocaleKey = 'en_US',
				ProfileId = communityProfileId
		);
		insert communityUser;
		System.runAs(communityUser)
		{
			List<User> users = new UsersSelector().elevatedSelectByEmployeeNumber(new Set<String>{'123'});
		}
	}
}
```

And this is the **Selector** class:
```
public with sharing class UsersSelector extends fflib_SObjectSelector 
{
	public List<Schema.SObjectField> getSObjectFieldList()
	{
		return new List<Schema.SObjectField>
		{
				User.Id,
				User.Username,
				User.EmployeeNumber
		};
	}

	public Schema.SObjectType getSObjectType()
	{
		return User.SObjectType;
	}

	public List<User> elevatedSelectByEmployeeNumber(Set<String> employeenumbers)
	{
		ElevatedContext elevatedContext = new ElevatedContext(this);
		return  (List<User>) elevatedContext.selectByEmployeeNumber(employeenumbers);
	}

	private without sharing class ElevatedContext extends fflib_SObjectSelector
	{
		UsersSelector selector;

		public ElevatedContext(UsersSelector selector) {
			super(true, false, false, true);
			this.selector = selector;
		}

		public List<Schema.SObjectField> getSObjectFieldList() {
			return selector.getSObjectFieldList();
		}

		public Schema.SObjectType getSObjectType() {
			return selector.getSObjectType();
		}

		public List<User> selectByEmployeeNumber(Set<String> employeenumbers) {
			return (List<User>)
					Database.query(
							newQueryFactory()
									.selectField('Profile.Name')
									.setCondition('EmployeeNumber IN :employeenumbers')
									.toSOQL()
					);
		}
	}
}
```

The unit test will fail with the following error:
```
FATAL_ERROR|System.ListException: List index out of bounds: 0
Class.fflib_QueryFactory.getFieldPath: line 111, column 1
Class.fflib_QueryFactory.selectField: line 207, column 1
Class.UsersSelector.ElevatedContext.selectByEmployeeNumber: line 56, column 1
Class.UsersSelector.elevatedSelectByEmployeeNumber: line 28, column 1
Class.fflib_SObjectSelectorTest.testElevatedSelector: line 175, column 1
```

And it fails on this line in the fflib_QueryFactory class:
```
lastSObjectType = tokenDescribe.getReferenceTo()[0]; //if it's polymorphic doesn't matter which one we get
```
It seem that the getReferenceTo method is not available on the DescribeFieldResult object.

The getFieldPath method is also using the fflib_SObjectDescribe method, and I was thinking that the caching of that class could also result in this issue, as the initial cached data might have been set in a "With Sharing" context while it was used in a "Without Sharing" context.

@afawcett Can you explain what the purpose if of this getFieldPath method? Because in my effort to resolve this issue I changed the getFieldPath method into the following and all seems to work fine again. But why do we have this method then?
```
	private String getFieldPath(String fieldName) {
		return fieldname;
	}
```

**Comment by JAertgeerts** [comment]
@afawcett @sfryffdc @stephenwillcock @pwrightffdc Is there a community owner within FinancialForce that actively looks at requests? Could you consider adding people from the community to help maintain this repo?

**Comment by daveespo** [comment]
I can't address the question of getting Pull Requests approved an merged, but I can comment on the purpose of getFieldPath()

getFieldPath() optionally enforces CRUD/FLS for the fields/relationships you specify in your Selector

List index out of bounds sounds like a bug in that method -- you probably don't have CRUD on the Lookup you're trying to traverse. I agree that it's peculiar that you'd encounter this but User is a Setup object so you might be running afoul of special rules with sharing for Setup objects

**Comment by wimvelzeboer** [comment]
Hi @daveespo 
Thanks for letting me know the purpose of the getFieldPath() methods, I did not realise that before.
The thing is this doesn't happen only on the User object but for all SObject types. I think it has something to do with the fflib_SObjectDescribe that is is caching the result, while it should reset it when the query is generated in a without sharing context.

I will try to do some testing with disabling the cache in fflib_SObjectDescribe and see it that resolves the issue.

**Comment by afawcett** [comment]
@JAertgeerts this project has now been moved to community owned now. Thanks for your patience. Will be working with other contributors on this repo to start work on open items. 👍

**Comment by foxysolutions** [comment]
The cause of this is the user not having access to that particular object (by Setup rules, e.g. a Guest User not having access to PermissionsetAssignment as [this is only allowed for Users who can assign Permissions](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_permissionsetassignment.htm)). Due to this the SObjectDescribe is not complete, causing a.o. the getReferenceTo() to return an empty list.

In my current projects I've implemented a fix, but there are several other additional features like allowing Casting for Polymorph Lookups and processing Aggregated SOQL functions (see PR #258 ). As I'm unfortunately not having the time to adopt to SFDX and separate this per PR, I'd like to share the additional code snippet below, for everyone running into this issue. I'll sure will try to create a PR soon.

The logic of the fix is based only for Standard SF fields (where this issue occurs) and is based on the assumption that SF Fields always have the SObjectType as fieldname, which was correct for all our test cases.

As you can see this structure also allows to specify logic for Polymorph (changing the second else-if to only equal size of 1 and then in the else specify more logic) as we experienced some troubles on Case.Owner.Name getting the exception "Invalid field 'Name' on 'Group'", but that will be another PR.

In case of any questions on how to implement the below, feel free to reach out!

Code snippet which should be placed as replacement of the line:
`lastSObjectType = relatedSObjectList[ 0 ]; //if it's polymorphic doesn't matter which one we get`

````List<Schema.SObjectType> relatedSObjectList = tokenDescribe.getReferenceTo();
if( relatedSObjectList.isEmpty() ){
   // Sometimes running users have no Standard SF Object access due to Setup restrictions (e.g. PermissionsetAssignment only available for users with ManageUsers)
   // To allow continuation of Query construction, and only when it concerns a standard SF field, use the field name which always equals the SObjectType
   if( tokenDescribe.custom == false ){
      lastSObjectType = fflib_SObjectDescribe.getDescribe( field ).getSObjectType();
   } else{
      throw new InvalidFieldException( 'User has no access to get the SObjectDescribe for field '+ field );
   }
} else if( relatedSObjectList.size() >= 1 ){
   lastSObjectType = relatedSObjectList[ 0 ]; //if it's polymorphic doesn't matter which one we get
}````

**Comment by ctchipps** [comment]
@foxysolutions It looks like this is going to be fixed in API v51.0, as per the discussion in #309. 

> VERSIONED BEHAVIOR CHANGE: From version 51.0 onwards, the getReferenceTo() method in the Schema.DescribeFieldResult class returns referenced objects that aren’t accessible to the context user.
If the context user has access to an object’s field that references another object, irrespective of the context user’s access to the cross-referenced object, the DescribeFieldResult.getReferenceTo() method returns references. In version 50.0 and earlier, if the context user doesn’t have access to the cross-referenced object, the method returns an empty list.

**Comment by daveespo** [comment]
Per the comment on #309 and with Spring '21 and the #317 API upgrade, this should be resolved. Please reopen the issue if you find differently
